src/network/network_coordinator.h

```diff
@@ -0,0 +1,103 @@
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file network_coordinator.h Part of the network protocol handling Game Coordinator requests. */
+
+#ifndef NETWORK_COORDINATOR_H
+#define NETWORK_COORDINATOR_H
+
+#include "core/tcp_coordinator.h"
+#include "network_stun.h"
+#include "network_turn.h"
+#include <map>
+
+/**
+ * Game Coordinator communication.
+ *
+ * For servers:
+ *  - Server sends CLIENT_REGISTER.
+ *  - GC probes server to check if it can directly connect, needs STUN, etc.
+ *  - GC sends SERVER_REGISTER_ACK with type of connection.
+ *  - Server sends every 30 seconds CLIENT_UPDATE.
+ *
+ * For clients (listing):
+ *  - Client sends CLIENT_LISTING.
+ *  - GC returns the full list of public servers via SERVER_LISTING (multiple packets).
+ *
+ * For clients (connecting):
+ *  - Client sends CLIENT_CONNECT.
+ *  - GC checks what type of connections the servers supports:
+ *    1) Direct connect?
+ *        - Send the client a SERVER_CONNECT with the peer address.
+ *        - a) Client connects, closes GC connection.
+ *        - b) Client connect fails, client sends CLIENT_CONNECT_FAILED to GC.
+ *    2) STUN?
+ *        - GC sends SERVER_STUN_REQUEST to server/client (asking for both IPv4 and IPv6).
+ *        - GC collects what combination works and sends SERVER_STUN_CONNECT to server/client.
+ *        - a) Server/client connect, client closes GC connection.
+ *        - b) Server/client connect fails, connecting side sends CLIENT_CONNECT_FAILED to GC.
+ *        - GC tries other combination if available.
+ *    3) TURN?
+ *        - GC sends SERVER_TURN_CONNECT to server/client.
+ *        - a) Server/client connect, client closes GC connection.
+ *        - b) Server/client connect fails, both send CLIENT_CONNECT_FAILED to GC.
+ *  - If all fails, GC sends SERVER_CONNECT_FAILED to indicate no connection was possible.
+ */
+
+/** Class for handling the client side of the Game Coordinator connection. */
+class ClientNetworkCoordinatorSocketHandler : public NetworkCoordinatorSocketHandler {
+private:
+	std::chrono::steady_clock::time_point next_update; ///< When to send the next update (if server and public).
+	std::map<std::string, TCPServerConnecter *> connecter; ///< Based on tokens, the current connecters that are pending.
+	std::map<std::string, TCPServerConnecter *> connecter_pre; ///< Based on join-keys, the current connecters that are pending.
+	std::map<std::string, std::map<int, std::unique_ptr<ClientNetworkStunSocketHandler>>> stun_handlers; ///< All pending STUN handlers.
```

Comment:
I'm not an expert on how all STUN handlers are managed and they work, but wouldn't a `std::unordered_map` be better in this case? I mean, is there really much reason for them to be ordered?

Reply:
No, the order is not important. In fact, it always only walks all elements. But that is not what I picked `std::map`.

I simply looked at our code-base, and noticed we basically always use `std::map`, also where `std::unordered_map` wouldn't be an issue. Then I read online about the difference, and basically you have two camps. 1 yelling to use the latter as it is `O(1)`, the other saying that means little, as often the `O(1)` of `std::unordered_map` is slower than the `O(log N)` of `std::map`. Then I shrug, give up, and just do what was already done :P

Some parts of C++ are to iffy for me to have an opinion about. This is one of them. I leave it to people more knowledgeable than me about C++ to give a call on this :P

Reply:
Conceptually `std::unordered_map` would indeed be the better choice. However, since the number of elements is tiny (after all, it is for active connection attempts to servers), the benefits of `std::unordered_map` for lookups/insertions will probably not outweigh the extra cost for hashing and maintaining the hash table compared to a `std::map` with usually 0 or 1, or maybe 2 or 3 elements.
All in all it does not really matter in this case, and then keeping it "simple" might be the better choice.

Reply:
I'm sorry if this is total bullshit, but theoretically the `unordered_map` would usually be faster with many entries. So if we only have a couple of entries, sure, use `std::map` as it'll be quicker on average. And I couldn't see so many socket handlers being required so I guess the minor speed boost of `unordered_map` would be negligible?

![image](https://user-images.githubusercontent.com/43609023/123868538-20fc5f00-d930-11eb-9f49-5ff776309faa.png)